sdk/azidentity/chained_token_credential_test.go

```diff
@@ -153,6 +153,220 @@ func TestChainedTokenCredential_GetTokenWithUnavailableCredentialInChain(t *test
 	}
 }
 
+func TestChainedTokenCredential_ChecksThatSuccessfulCredentialIsSet(t *testing.T) {
+	err := initEnvironmentVarsForTest()
+	if err != nil {
+		t.Fatalf("Could not set environment variables for testing: %v", err)
+	}
+	srv, close := mock.NewTLSServer()
+	defer close()
+	srv.AppendResponse(mock.WithBody([]byte(accessTokenRespSuccess)))
+	options := ClientSecretCredentialOptions{}
+	options.AuthorityHost = AuthorityHost(srv.URL())
+	options.Transport = srv
+	secCred, err := NewClientSecretCredential(tenantID, clientID, secret, &options)
+	if err != nil {
+		t.Fatalf("Unable to create credential. Received: %v", err)
+	}
+	envCred, err := NewEnvironmentCredential(&EnvironmentCredentialOptions{
+		ClientOptions: azcore.ClientOptions{Transport: srv},
+		AuthorityHost: AuthorityHost(srv.URL()),
+	})
+	if err != nil {
+		t.Fatalf("Failed to create environment credential: %v", err)
+	}
+	cred, err := NewChainedTokenCredential([]azcore.TokenCredential{secCred, envCred}, nil)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	tk, err := cred.GetToken(context.Background(), policy.TokenRequestOptions{Scopes: []string{scope}})
+	if err != nil {
+		t.Fatalf("Received an error when attempting to get a token but expected none")
+	}
+	if tk.Token != tokenValue {
+		t.Fatalf("Received an incorrect access token")
+	}
+	if tk.ExpiresOn.IsZero() {
+		t.Fatalf("Received an incorrect time in the response")
+	}
+	if cred.successfulCredential == nil {
+		t.Fatalf("The successful credential was not assigned")
+	}
+	if cred.successfulCredential != secCred {
+		t.Fatalf("The successful credential should have been the secret credential")
+	}
+}
+
+/**
+ * Helps count the number of times a credential is called.
+ */
+type TestCountPolicy struct{ count int }
+
+/**
+ * Helps count the number of times a credential is called.
+ */
+func (p *TestCountPolicy) Do(req *policy.Request) (*http.Response, error) {
+	p.count += 1
+	return req.Next()
+}
+
+func TestChainedTokenCredential_RepeatedGetTokenWithSuccessfulCredential(t *testing.T) {
+	err := initEnvironmentVarsForTest()
+	if err != nil {
+		t.Fatalf("Could not set environment variables for testing: %v", err)
+	}
+	srv, close := mock.NewTLSServer()
+	defer close()
+
+	secretCountPolicy := &TestCountPolicy{}
+	environmentCountPolicy := &TestCountPolicy{}
+
+	srv.AppendResponse(mock.WithBody([]byte(accessTokenRespSuccess)))
+	srv.AppendResponse(mock.WithBody([]byte(accessTokenRespSuccess)))
+	options := ClientSecretCredentialOptions{}
+	options.AuthorityHost = AuthorityHost(srv.URL())
+	options.PerCallPolicies = []policy.Policy{secretCountPolicy}
+	options.Transport = srv
+	secCred, err := NewClientSecretCredential(tenantID, clientID, secret, &options)
```

Comment:
Suggest using fake credentials instead of real ones, to simplify the test and keep it focused on ChainedTokenCredential's behavior

Reply:
How does this look? https://github.com/Azure/azure-sdk-for-go/pull/16392/commits/20197a94527b5dc38f8d0270df23c0fccff47e68